app/uk/gov/hmrc/thirdpartyapplication/services/CredentialService.scala

```diff
@@ -103,6 +103,36 @@ class CredentialService @Inject()(applicationRepository: ApplicationRepository,
     } yield ApplicationTokenResponse(updatedApp.tokens.production)
   }
 
+  def deleteClientSecret(applicationId: UUID,
+                         clientSecretId: String,
+                         actorEmailAddress: String)(implicit hc: HeaderCarrier): Future[ApplicationTokenResponse] = {
+    def audit(applicationId: UUID, clientSecretId: String): Future[AuditResult] =
+      auditService.audit(ClientSecretRemoved, Map("applicationId" -> applicationId.toString, "removedClientSecret" -> clientSecretId))
+
+    def sendNotification(clientSecret: ClientSecret, app: ApplicationData): Future[HttpResponse] = {
+      emailConnector.sendRemovedClientSecretNotification(actorEmailAddress, clientSecret.name, app.name, app.admins.map(_.emailAddress))
+    }
+
+    def findClientSecretToDelete(application: ApplicationData, clientSecretId: String): ClientSecret = {
+      val clientSecretToUpdate = application.tokens.production.clientSecrets.find(_.id == clientSecretId)
+
+      if(clientSecretToUpdate.isEmpty) {
```

Comment:
I guess this could be one of those getOrElse(throw ...) but it's fine

Reply:
That's nicer, I'll update it to use that